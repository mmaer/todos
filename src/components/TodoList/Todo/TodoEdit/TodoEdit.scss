.todo-editing {
  display: block;
  position: relative;
  margin: 0 0 0 43px;
  width: 506px;
  font-size: 24px;
  font-family: inherit;
  font-weight: inherit;
  line-height: 1.4em;
  border: 0;
  color: inherit;
  padding: 12px 16px;
  border: 1px solid #999;
  box-shadow: inset 0 -1px 5px 0 rgba(0, 0, 0, 0.2);
  box-sizing: border-box;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}
