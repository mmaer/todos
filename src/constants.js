export const LOCAL_STORAGE_KEY = "todos";

export const ENTER_KEY = 13;
export const ESCAPE_KEY = 27;

export const FILTER_ALL = "all";
export const FILTER_ACTIVE = "active";
export const FILTER_COMPLETED = "completed";

export const FILTERS = [FILTER_ALL, FILTER_ACTIVE, FILTER_COMPLETED];
