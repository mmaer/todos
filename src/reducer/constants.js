export const ADD_TODO = "ADD_TODO";
export const REMOVE_TODO = "REMOVE_TODO";
export const EDIT_TODO_TITLE = "EDIT_TODO_TITLE";

export const TOGGLE_COMPLETED_STATUS = "TOGGLE_COMPLETED_STATUS";
export const TOGGLE_EDITING_STATUS = "TOGGLE_EDITING_STATUS";

export const REMOVE_COMPLETED_TODOS = "REMOVE_COMPLETED_TODOS";
