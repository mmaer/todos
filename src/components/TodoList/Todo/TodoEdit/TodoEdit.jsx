import React, { useState, memo } from "react";

import { ENTER_KEY, ESCAPE_KEY } from "../../../../constants";

import "./TodoEdit.scss";

const TodoEdit = memo(
  ({ title, editing, id, removeTodo, editTodoTitle, toggleEditingStatus }) => {
    const [value, setValue] = useState(title);

    const handleOnChange = ({ target }) => {
      setValue(target.value);
    };

    const save = () => {
      if (editing) {
        if (value === "") {
          removeTodo(id);
        } else {
          editTodoTitle(id, value);
          toggleEditingStatus(id);
        }
      }
    };

    const handleKeys = event => {
      event.persist();
      if (event.keyCode === ENTER_KEY) {
        save();
      } else if (event.keyCode === ESCAPE_KEY) {
        setValue(title);
        toggleEditingStatus(id);
      }
    };

    return (
      <input
        className="todo-editing"
        onKeyUp={handleKeys}
        onBlur={save}
        value={value}
        onChange={handleOnChange}
      />
    );
  }
);

export default TodoEdit;
