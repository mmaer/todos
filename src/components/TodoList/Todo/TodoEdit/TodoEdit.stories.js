import React from "react";

import TodoEdit from ".";

export default {
  title: "TodoEdit",
};

const TODO = { title: "Todo", editing: true, id: 1 };

export const todoEdit = () => <TodoEdit {...TODO} />;
