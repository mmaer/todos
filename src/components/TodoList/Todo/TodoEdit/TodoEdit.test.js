import React from "react";
import { mount } from "enzyme";

import { ENTER_KEY, ESCAPE_KEY } from "../../../../constants";

import TodoEdit from "./";

describe("TodoEdit component", () => {
  let mockEditTodoTitleFunc;
  let mockToggleEditingStatusFunc;
  let mockRemoveTodoFunc;
  let wrapper;

  const TODO = {
    title: "title",
    completed: false,
    editing: true,
    id: 1,
    removeTodo: () => {},
    editTodoTitle: () => {},
    toggleEditingStatus: () => {},
  };

  const newTitle = "new-title";

  const simulate = (wrapper, value, keyCode) => {
    wrapper.simulate("change", { target: { value } });
    wrapper.simulate("keyup", { persist: () => {}, keyCode });
  };

  beforeEach(() => {
    mockEditTodoTitleFunc = jest.fn();
    mockToggleEditingStatusFunc = jest.fn();
    mockRemoveTodoFunc = jest.fn();

    wrapper = mount(
      <TodoEdit
        {...TODO}
        editTodoTitle={mockEditTodoTitleFunc}
        toggleEditingStatus={mockToggleEditingStatusFunc}
        removeTodo={mockRemoveTodoFunc}
      />
    );
  });

  it("Should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("Should call editTodoTitle and toggleEditingStatus functions when new value is passed and enter key is pressed", () => {
    simulate(wrapper, newTitle, ENTER_KEY);

    expect(mockEditTodoTitleFunc).toBeCalledWith(TODO.id, newTitle);
    expect(mockToggleEditingStatusFunc).toBeCalledWith(TODO.id);
  });

  it("Should call removeTodo function when value is empty and enter key is pressed", () => {
    simulate(wrapper, "", ENTER_KEY);

    expect(mockRemoveTodoFunc).toBeCalledWith(TODO.id);
  });

  it("Should call toggleEditingStatus function when escape key is pressed", () => {
    simulate(wrapper, newTitle, ESCAPE_KEY);

    expect(mockToggleEditingStatusFunc).toBeCalledWith(TODO.id);
    expect(wrapper.find("input").props().value).toEqual(TODO.title);
  });

  it("Should do nothing when editing prop is equal false", () => {
    wrapper.setProps({ editing: false });
    simulate(wrapper, newTitle, ENTER_KEY);

    expect(mockEditTodoTitleFunc.mock.calls.length).toBe(0);
    expect(mockToggleEditingStatusFunc.mock.calls.length).toBe(0);
    expect(mockRemoveTodoFunc.mock.calls.length).toBe(0);
  });
});
