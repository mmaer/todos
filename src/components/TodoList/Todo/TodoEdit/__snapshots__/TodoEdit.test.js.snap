// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoEdit component Should match snapshot 1`] = `
<Memo()
  completed={false}
  editTodoTitle={[MockFunction]}
  editing={true}
  id={1}
  removeTodo={[MockFunction]}
  title="title"
  toggleEditingStatus={[MockFunction]}
>
  <input
    autoFocus={true}
    className="todo-editing"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyUp={[Function]}
    value="title"
  />
</Memo()>
`;
